## Microsoft Industry Project

Microsoft has been a leader in the tech industry for many years due to its prodcut development and technological advancements. With their revenue and profits soaring throughout the years, they have decided to venture into different industries in order to diversify their investments. They have hired 4 data scientists to explore and analyze the movie industry data, and set recommendations for their first movie.

### Questions to be answered

1- What are the  top 5 genres with the highest ROIs

2- What were the production costs for the top 5 genres with the highest ROIs

3- Do release date months affect the ROI of the 5 genres

## Importing the libraries

In [1]:
# import python libraries
import seaborn as sns
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd

## Reading the data

In [2]:
# open the first data file
title_basics = pd.read_csv('Data/imdb.title.basics.csv.gz')
title_basics.head()
title_basics['start_year'].min()

2010

In [3]:
# open the second data file
title_ratings = pd.read_csv('Data/imdb.title.ratings.csv.gz')
title_ratings.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [4]:
# open the third data file
movie_budgets = pd.read_csv('Data/tn.movie_budgets.csv.gz')
movie_budgets.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747

## Data Preperation
Before performing any analysis, the data has to be cleaned and merged for a proper dataset

***Steps to preare data for title_basics, (-)= completed***
1. Rename and drop columns
2. Calculate percentage of nulls in genre and drop null values

In [5]:
title_basics.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [6]:
# rename columns for future merging
title_basics.rename(columns = {'primary_title':'movie'}, inplace = True)

In [7]:
# drop columns not being used for analysis
title_basics.drop(columns = ["original_title","runtime_minutes"], inplace=True)

***Steps to preare data for title_basics***
1. Rename and drop columns(-) 
2. Calculate percentage of nulls in genre and drop null values

In [8]:
title_basics['genres'].isna().sum()/146144*100

3.7004598204510617

In [9]:
# drop null values for genres
title_basics.dropna(subset= ['genres'], inplace = True)

In [10]:
title_basics.head()

tconst                            movie  start_year  \
0  tt0063540                        Sunghursh        2013   
1  tt0066787  One Day Before the Rainy Season        2019   
2  tt0069049       The Other Side of the Wind        2018   
3  tt0069204                  Sabse Bada Sukh        2018   
4  tt0100275         The Wandering Soap Opera        2017   

                 genres  
0    Action,Crime,Drama  
1       Biography,Drama  
2                 Drama  
3          Comedy,Drama  
4  Comedy,Drama,Fantasy

***Steps to prepare data for title_basics***
1. Rename and drop columns(-) 
2. Calculate percentage of nulls in genre and drop null values(-)

***

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time
2. Change production_budget, domestic_gross, and worldwide_budget to integers
3. Drop id column and drop rows where worldwide_gross or production_budget == 0
4. Add profit column & Calculate:
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places
6. Add movie year column

In [11]:
# change release date to date-time
movie_budgets['release_date'] = pd.to_datetime(movie_budgets['release_date'])

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers
3. Drop id column and drop rows where worldwide_gross or production_budget == 0
4. Add profit column & Calculate:
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places
6. Add movie year column

In [12]:
# change production_budget, domestic_gross, and worldwide_budget to integers
movie_budgets['production_budget'] = movie_budgets['production_budget'].replace('[\$,]', '', regex=True).astype(float)
movie_budgets['domestic_gross'] = movie_budgets['domestic_gross'].replace('[\$,]', '', regex=True).astype(float)
movie_budgets['worldwide_gross'] = movie_budgets['worldwide_gross'].replace('[\$,]', '', regex=True).astype(float)

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers(-)
3. Drop id column and drop rows where worldwide_gross or production_budget == 0
4. Add profit column & Calculate:
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places
6. Add movie year column

In [13]:
# drop id column
movie_budgets.drop(columns = ["id"], inplace=True)
# drop rows where worldwide_gross or production_budget is 0
movie_budgets = movie_budgets[movie_budgets.worldwide_gross != 0]
movie_budgets = movie_budgets[movie_budgets.production_budget != 0]

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers(-)
3. Drop id column and drop rows where worldwide_gross or production_budget == 0(-)
4. Add profit column & Calculate:
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places
6. Add movie year column

In [14]:
# add a profit column to movie_budgets
movie_budgets['profit'] = movie_budgets['worldwide_gross'] - movie_budgets['production_budget']
# calculate profit in millions
movie_budgets['profit_million']=(movie_budgets['profit']/1000000)
# calculate production budget in millions
movie_budgets['production_budget_million']=(movie_budgets['production_budget']/1000000)

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers(-)
3. Drop id column and drop rows where worldwide_gross or production_budget == 0(-)
4. Add profit column & Calculate:(-)
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places
6. Add movie year column

In [15]:
# add a column ROI calculating the return on investment
movie_budgets['ROI'] = (movie_budgets['profit'] / movie_budgets['production_budget']) * 100
# round ROI to 2 decimal places
movie_budgets = movie_budgets.round({'ROI': 2})

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers(-)
3. Drop id column and drop rows where worldwide_gross or production_budget == 0(-)
4. Add profit column & Calculate:(-)
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places(-)
6. Add movie year column

In [16]:
# add a column for movie year
movie_budgets['year'] = pd.DatetimeIndex(movie_budgets['release_date']).year

***Steps to prepare data for movie_budgets***
1. Change column release_date to date-time(-)
2. Change production_budget, domestic_gross, and worldwide_budget to integers(-)
3. Drop id column and drop rows where worldwide_gross or production_budget == 0(-)
4. Add profit column & Calculate:(-)
    1. profit in millions
    2. production budget in millions
5. Add ROI column(calculates return on investment) and round to 2 decimal places(-)
6. Add movie year column(-)

In [17]:
movie_budgets.head()

release_date                                        movie  \
0   2009-12-18                                       Avatar   
1   2011-05-20  Pirates of the Caribbean: On Stranger Tides   
2   2019-06-07                                 Dark Phoenix   
3   2015-05-01                      Avengers: Age of Ultron   
4   2017-12-15            Star Wars Ep. VIII: The Last Jedi   

   production_budget  domestic_gross  worldwide_gross        profit  \
0        425000000.0     760507625.0     2.776345e+09  2.351345e+09   
1        410600000.0     241063875.0     1.045664e+09  6.350639e+08   
2        350000000.0      42762350.0     1.497624e+08 -2.002376e+08   
3        330600000.0     459005868.0     1.403014e+09  1.072414e+09   
4        317000000.0     620181382.0     1.316722e+09  9.997217e+08   

   profit_million  production_budget_million     ROI  year  
0     2351.345279                      425.0  553.26  2009  
1      635.063875                      410.6  154.67  2011  
2     -200.237650                      350.0  -57.21  2019  
3     1072.413963                      330.6  324.38  2015  
4      999.721747                      317.0  315.37  2017

***

## Merging Data Tables
Description of what we are doing: 'Merge the two tables in order to get a proper dataframe' - review 

***Steps to merge data for new dataframe: mdf, (-)= completed***
1. Merge title_basics and title_ratings into: imdb_data
2. Merge imdb_data and movie_budgets into: mdf

In [18]:
# merge title_basics and title_ratings on tconst
imdb_data = pd.merge(title_basics, title_ratings, how = 'inner', on = 'tconst')

In [19]:
# merge imdb_data with movie_budgets
mdf = pd.merge(imdb_data, movie_budgets, how = 'inner', on = 'movie')

***Steps to merge data for new dataframe: mdf, (-)= completed***
1. Merge title_basics and title_ratings into: imdb_data(-)
2. Merge imdb_data and movie_budgets into: mdf(-)

In [20]:
mdf.shape

(2592, 15)

***

## Preparing mdf dataframe
The dataframe ***mdf*** needs to be cleaned, sorted and filtered in order to keep the proper values

***Steps to prepare dataframe: mdf, (-)= completed***
1. Move the movies with the most votes on imdb_data up to top
2. Create a column same_year that is 0 if the movies have the same start year and same year from the release date column
3. Change mdf to include movies only if same_year == 0
4. Drop:
    1. duplicate titles
    2. tconst

In [21]:
#1
mdf=mdf.sort_values(by='numvotes', ascending = False)

***Steps to prepare dataframe: mdf, (-)= completed***
1. Move the movies with the most votes on imdb_data up to top(-)
2. Create a column same_year that is 0 if the movies have the same start year and same year from the release date column
3. Change mdf to include movies only if same_year == 0
4. Drop:
    1. duplicate titles
    2. tconst

In [22]:
#2
mdf['same_year']= mdf['start_year']-mdf['year']

In [23]:
#3
mdf = mdf.loc[mdf['same_year']== 0]

***Steps to prepare dataframe: mdf, (-)= completed***
1. Move the movies with the most votes on imdb_data up to top(-)
2. Create a column same_year that is 0 if the movies have the same start year and same year from the release date column(-)
3. Change mdf to include movies only if same_year == 0(-)
4. Drop:
    1. duplicate titles
    2. tconst

In [24]:
#4
mdf = mdf.drop_duplicates(subset='movie')
mdf.drop(columns = ["tconst"], inplace=True)

***Steps to prepare dataframe: mdf, (-)= completed***
1. Move the movies with the most votes on imdb_data up to top(-)
2. Create a column same_year that is 0 if the movies have the same start year and same year from the release date column(-)
3. Change mdf to include movies only if same_year == 0(-)
4. Drop:(-)
    1. duplicate titles
    2. tconst

In [25]:
mdf.shape

(1357, 15)

***

## Checking for data rows
The IMDB data contains movies only from 2010 on. The movie budgets table has movies ranging back to 1915, so we figured that when we merge the two tables we will be limited in the number of movies that are shared between the two tables. 

To make sure our merge of IMDB data with the movie budgets provided us with a good number of data, before merging we looked at the mean ROI from the movie budgets table for the years after 2010.  Then we compared it to the mean ROI after merging the two tables.

***Steps to check for number of data rows, (-)= completed***
1. Create movie_budgets_filtered, filtering the values from movie_budgets['year'] >= 10
2. Find the ROI mean of movie_budgets_filtered
3. Find the ROI mean of mdf

In [26]:
#1
movie_budgets_filtered = movie_budgets[movie_budgets['year'] >= 2010]

In [27]:
#2
movie_budgets_filtered['ROI'].mean()

266.4769458896981

In [28]:
#3
mdf['ROI'].mean()

298.7207295504783

The mean ROI for the data before merging is slightly lower than the data after merging, however they are close enough to continue working with the reduced amount of data.

## Finding how to subset mdf['genre']
We then explored how to proceed with the gernes. First we investigated whether a subset of the  combined genres would represent enough of the movie data to look at the combined genres as individual categories.

In [29]:
mdf['genres'].value_counts()

Adventure,Animation,Comedy      67
Action,Adventure,Sci-Fi         51
Comedy,Drama,Romance            49
Comedy                          48
Comedy,Drama                    42
                                ..
Drama,Music,Musical              1
Adventure,Comedy,Romance         1
Action,Comedy,Sport              1
Adventure,Drama,History          1
Action,Biography,Documentary     1
Name: genres, Length: 214, dtype: int64

There are 214 unique genre combinations. We look at the top 20 genres (about 10% of the total genres) to see if that represented enough of the data set to proceed with the combined genres.

In [34]:
#ask how this code works*
mdf_combined_genres = mdf[mdf['genres'].map(mdf['genres'].value_counts() >= 17)] 

In [35]:
mdf_combined_genres['genres'].value_counts()

Adventure,Animation,Comedy    67
Action,Adventure,Sci-Fi       51
Comedy,Drama,Romance          49
Comedy                        48
Comedy,Drama                  42
Drama                         40
Comedy,Romance                40
Action,Crime,Drama            38
Drama,Romance                 32
Action,Adventure,Fantasy      31
Action,Comedy,Crime           29
Horror,Mystery,Thriller       29
Action,Adventure,Drama        26
Drama,Thriller                21
Crime,Drama,Thriller          21
Action,Crime,Thriller         21
Biography,Drama,History       20
Action,Adventure,Comedy       20
Horror,Thriller               20
Biography,Comedy,Drama        17
Name: genres, dtype: int64

In [ ]:
mdf_combined_genres['genres'].value_counts().sum()

In [ ]:
662/1357

The top 20 combined genres, which is about 10% of the combined genres, represent less than 50% of the movies in the data. For that reason, it will be more helpful to look at the movies separated into single genres.

## Seperating mdf['genre'] 
Each genre is seperated into its own value in order to compare the averages of the ROI. This will allow us to view the genres with the highest average ROI.

***Steps to seperating genres, (-)= completed***
1. Seperate genres column values into lists
2. Create a new dataframe mdf2 to filter each genre within a movie
3. Remove 'The Gallows' movie: is an outlier
4. Create sorted_genres2 dataframe which is grouped by the unique genres

In [37]:
#1
mdf["genres"] = mdf["genres"].apply(lambda x: x.split(",") if type(x) == str else x)

In [38]:
mdf.shape

(1357, 15)

In [39]:
mdf.head()

movie  start_year                       genres  \
510               Inception        2010  [Action, Adventure, Sci-Fi]   
477   The Dark Knight Rises        2012           [Action, Thriller]   
96             Interstellar        2014   [Adventure, Drama, Sci-Fi]   
1296       Django Unchained        2012             [Drama, Western]   
110            The Avengers        2012  [Action, Adventure, Sci-Fi]   

      averagerating  numvotes release_date  production_budget  domestic_gross  \
510             8.8   1841066   2010-07-16        160000000.0     292576195.0   
477             8.4   1387769   2012-07-20        275000000.0     448139099.0   
96              8.6   1299334   2014-11-05        165000000.0     188017894.0   
1296            8.4   1211405   2012-12-25        100000000.0     162805434.0   
110             8.1   1183655   2012-05-04        225000000.0     623279547.0   

      worldwide_gross        profit  profit_million  \
510      8.355246e+08  6.755246e+08      675.524642   
477      1.084439e+09  8.094391e+08      809.439099   
96       6.663794e+08  5.013794e+08      501.379375   
1296     4.499483e+08  3.499483e+08      349.948323   
110      1.517936e+09  1.292936e+09     1292.935897   

      production_budget_million     ROI  year  same_year  
510                       160.0  422.20  2010          0  
477                       275.0  294.34  2012          0  
96                        165.0  303.87  2014          0  
1296                      100.0  349.95  2012          0  
110                       225.0  574.64  2012          0

In [42]:
#2
mdf2 = mdf.explode('genres')

In [43]:
mdf2['genres'].unique()

array(['Action', 'Adventure', 'Sci-Fi', 'Thriller', 'Drama', 'Western',
       'Biography', 'Crime', 'Mystery', 'Comedy', 'Family', 'Fantasy',
       'Animation', 'Romance', 'Music', 'History', 'Horror', 'Sport',
       'War', 'Musical', 'Documentary'], dtype=object)

In [44]:
len(mdf2['genres'].unique())

21

In [45]:
#3
no_gallows = mdf2.drop(index = 1722).sort_values(by="ROI",ascending = False)

In [46]:
#4
sorted_genres2 = no_gallows.groupby(["genres"], as_index=False).mean().sort_values(by = "ROI", ascending=False)

top5genres_v2 = list(sorted_genres2['genres'].values[0:5])
top5genres_v2

['Horror', 'Mystery', 'Thriller', 'Sci-Fi', 'Animation']

In [47]:
sorted_genres2

genres   start_year  averagerating       numvotes  production_budget  \
11       Horror  2014.167785       5.708054   77185.268456       2.051299e+07   
14      Mystery  2014.200000       6.190435  135118.060870       2.644637e+07   
18     Thriller  2013.753247       6.129004  133254.857143       3.568663e+07   
16       Sci-Fi  2014.388889       6.565873  286807.888889       9.956806e+07   
2     Animation  2014.193878       6.574490  128583.132653       1.001939e+08   
15      Romance  2013.072626       6.338547   82429.480447       2.310958e+07   
12        Music  2013.723404       6.389362   69891.170213       1.812872e+07   
3     Biography  2014.784615       7.051538  113389.415385       2.651900e+07   
7         Drama  2013.860360       6.596396  106750.448949       2.838700e+07   
4        Comedy  2013.627083       6.240625   98136.629167       4.240746e+07   
9       Fantasy  2013.905172       6.128448  155112.741379       8.758922e+07   
1     Adventure  2014.244118       6.487059  204891.002941       1.090544e+08   
6   Documentary  2012.703704       6.655556    9978.370370       9.303630e+06   
13      Musical  2013.000000       6.171429   87704.857143       5.498571e+07   
8        Family  2013.400000       6.187059   84257.682353       7.303412e+07   
0        Action  2014.043165       6.366906  187935.441247       8.380576e+07   
10      History  2015.051282       6.946154  100668.974359       3.558974e+07   
5         Crime  2013.930876       6.417972  121316.880184       3.379041e+07   
17        Sport  2013.937500       6.931250   74843.875000       2.550625e+07   
20      Western  2012.444444       6.766667  217706.555556       6.342222e+07   
19          War  2013.588235       6.570588   73988.294118       2.826471e+07   

    domestic_gross  worldwide_gross        profit  profit_million  \
11    3.630081e+07     7.965158e+07  5.913859e+07       59.138593   
14    4.246262e+07     9.798276e+07  7.153639e+07       71.536385   
18    4.924466e+07     1.279535e+08  9.226690e+07       92.266902   
16    1.329984e+08     3.647116e+08  2.651435e+08      265.143504   
2     1.407021e+08     3.887555e+08  2.885616e+08      288.561604   
15    3.410309e+07     7.047518e+07  4.736560e+07       47.365597   
12    3.564235e+07     7.776171e+07  5.963299e+07       59.632989   
3     4.041363e+07     8.309290e+07  5.657390e+07       56.573903   
7     3.581019e+07     7.877166e+07  5.038466e+07       50.384662   
4     6.237079e+07     1.424374e+08  1.000299e+08      100.029917   
9     8.472140e+07     2.476530e+08  1.600638e+08      160.063750   
1     1.278140e+08     3.614190e+08  2.523646e+08      252.364567   
6     1.341404e+07     2.864313e+07  1.933950e+07       19.339502   
13    1.086679e+08     2.641159e+08  2.091302e+08      209.130224   
8     8.647824e+07     2.135799e+08  1.405458e+08      140.545807   
0     9.286873e+07     2.564038e+08  1.725980e+08      172.598045   
10    4.192146e+07     8.782263e+07  5.223288e+07       52.232883   
5     3.873408e+07     8.622434e+07  5.243393e+07       52.433930   
17    3.490760e+07     5.469688e+07  2.919063e+07       29.190626   
20    5.900548e+07     1.279120e+08  6.448982e+07       64.489819   
19    2.415504e+07     5.778184e+07  2.951713e+07       29.517132   

    production_budget_million         ROI         year  same_year  
11                  20.512987  698.902752  2014.167785        0.0  
14                  26.446371  598.706087  2014.200000        0.0  
18                  35.686635  419.959437  2013.753247        0.0  
16                  99.568056  283.657778  2014.388889        0.0  
2                  100.193878  278.690204  2014.193878        0.0  
15                  23.109581  267.470168  2013.072626        0.0  
12                  18.128723  261.097021  2013.723404        0.0  
3                   26.519000  248.377846  2014.784615        0.0  
7                   28.386996  238.584444  2013.860360        0.0  
4                   42.4074